src/backend/access/appendonly/appendonlyam.c

```diff
@@ -148,7 +148,7 @@ initscan(AppendOnlyScanDesc scan, ScanKey key)
 	/*
 	 * copy the scan key, if appropriate
 	 */
-	if (key != NULL)
+	if (key != NULL && scan->aos_key != NULL && scan->aos_nkeys > 0)
```

Comment:
```suggestion
	if (key != NULL && scan->aos_nkeys > 0)
```
Maybe remove extra check because the only place where `scan->aos_nkeys = nkeys` is set is at https://github.com/arenadata/gpdb/blob/3b5aa3e9d5141832d94172e20e41a75770c07f34/src/backend/access/appendonly/appendonlyam.c#L1590 and then the count is checked at https://github.com/arenadata/gpdb/blob/3b5aa3e9d5141832d94172e20e41a75770c07f34/src/backend/access/appendonly/appendonlyam.c#L1639-L1642?

Reply:
> the only place where `scan->aos_nkeys = nkeys` is set is at

how about appendonly_rescan?

Reply:
I didn't find any other places where `aos_nkeys` is changed.

Reply:
why not leave
```suggestion
	if (key != NULL && scan->aos_key != NULL)
```
?

Reply:
I think the condition with `aos_nkeys` shows the expected behavior better. Maybe add an assert before in the condition to silence the static analyzer? I'm okay with the original change as well, but it may confuse future maintainers :(

Reply:
I think the NULL check is to prevent UB for memcpy, and the length check is to prevent an unnecessary call.

Reply:
> an assert

does not

> silence the static analyzer